### PR TITLE
Define Webhook payload model and store those on the DB

### PR DESCRIPTION
**Pull request by leite08**
Ticket: #43

### Dependencies

- Upstream: none
- Downstream: https://github.com/metriport/metriport-internal/pull/148

### Description

Define the Webhook (WH) payload format and  add a table to store the WH requests on the DB:
- Determine WH payload model
- Converts Metriport Data model into the WH payload model
- Creates the `webhook_request` table to store WH requests on the DB
- Also:
  - fix table creation on migrations, it should not refer to the current model, as that evolves over time, and the migration is a snapshot
  - enable access to `createdAt` and `updatedAt` on models
  - fix date types on Sequelize models to be `DATE` and not `STRING`
  - fix bug on `GET /user/providers`, it failed when user was not connected to any Provider

### Release Plan

- [X] ⚠️Contains a DB migration ⚠️
- Plan
  - [ ] Release this
  - [ ] Release downstream dependencies

### CODE REVIEW

File: api/app/src/command/webhook/webhook-request.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { v4 as uuidv4 } from "uuid";
+import NotFoundError from "../../errors/not-found";
+import {
+  WebhookRequest,
+  WebhookRequestStatus,
+} from "../../models/webhook-request";
+
+export type CreateWebhookRequestCommand = { cxId: string; payload: unknown };
```
Comment (Goncharo) [inline api/app/src/command/webhook/webhook-request.ts:8]: didn't know `unknown` was a thing, cool!

File: api/app/src/command/webhook/webhook.ts [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+type WebhookUserDataPayload = {
+  [k in DataType]?: MetriportData[];
+};
+type WebhookUserPayload = { userId: string } & WebhookUserDataPayload;
+type WebhookMetadataPayload = { messageId: string; when: string };
+type WebhookDataPayload = {
+  meta: WebhookMetadataPayload;
+  users: WebhookUserPayload[];
+};
+type WebhookDataPayloadWithoutMessageId = Omit<WebhookDataPayload, "meta">;
```
Comment (Goncharo) [inline api/app/src/command/webhook/webhook.ts:32]: first time seeing `Omit` as well, fancy

File: api/app/src/command/webhook/webhook.ts [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+  return false;
+};
+
+const sendPayload = async (
+  payload: WebhookPayload,
+  url: string,
+  apiKey: string,
+  timeout = 5_000
+): Promise<any> => {
+  log(`Sending payload:\n${JSON.stringify(payload)}`);
```
Comment (Goncharo) [inline api/app/src/command/webhook/webhook.ts:223]: Is this for debugging purposes? Can blow up our logs fairly quickly

File: api/app/src/models/webhook-request.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        },
+        payload: {
+          type: DataTypes.JSONB,
+        },
+        status: {
+          type: DataTypes.STRING,
+        },
+        statusAt: {
+          type: DataTypes.DATE(6),
+        },
```
Comment (Goncharo) [inline api/app/src/models/webhook-request.ts:32]: Looks like we only use this during create(), if that's the case, can't we just use the default createdAt column for this?

File: connect-widget/app/src/pages/connect/components/connect-button.tsx [hunk 5]
```diff
@@ -25,15 +25,15 @@ const ConnectButton = ({
...
+  // const garmin = useBreakpointValue({
+  //   base: <Text p="4">+</Text>,
+  //   sm: <Text p="4">Coming Soon!</Text>,
+  // });
 
   return (
     <Box>
       <Center>
-        {provider.name === "garmin" ? (
+        {/* {provider.name === "garmin" ? (
```
Comment (Goncharo) [inline connect-widget/app/src/pages/connect/components/connect-button.tsx:36]: I assume the comments are here on purpose so we can reference this in the future?

File: api/app/src/command/webhook/webhook.ts [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    return true;
+  } catch (err: any) {
+    try {
+      // mark this request as failed on the DB
+      await updateWebhookRequestStatus({
+        id: webhookRequest.id,
+        status: "failure",
+      });
+    } catch (err2) {
+      console.log(`Failed to store failure state on WH log`, err);
```
Comment (Goncharo) [inline api/app/src/command/webhook/webhook.ts:194]: `err2`?

File: api/app/src/command/webhook/webhook.ts [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+    }
+    try {
+      // update the status of this webhook on the DB
+      await updateWebhookStatus({
+        id: settings.id,
+        webhookEnabled: false,
+        webhookStatusDetail,
+      });
+    } catch (err2) {
+      console.log(`Failed to store failure state on WH settings`, err);
```
Comment (Goncharo) [inline api/app/src/command/webhook/webhook.ts:211]: `err2`?

**Review by Goncharo — APPROVED** [review]
Minor comments/questions, generally looks good!

File: api/app/src/models/webhook-request.ts (hunk 4 above)
Comment (leite08) [reply api/app/src/models/webhook-request.ts:32]: the status can be updated as we try again, that's why I thought about a diff column... as for why not `updatedAt`, in theory they should be the same, as the only thing that should be updatable on a record in this table is the status, maybe we can use that then?

File: connect-widget/app/src/pages/connect/components/connect-button.tsx (hunk 5 above)
Comment (leite08) [reply connect-widget/app/src/pages/connect/components/connect-button.tsx:36]: This was commented out in another PR where we temporarily disabled Garmin, not sure why it's showing as difference... but yeah, we'll reenable this soon.

File: api/app/src/command/webhook/webhook.ts (hunk 6 above)
Comment (leite08) [reply api/app/src/command/webhook/webhook.ts:194]: good catch!

File: api/app/src/command/webhook/webhook.ts (hunk 3 above)
Comment (leite08) [reply api/app/src/command/webhook/webhook.ts:223]: oh yeah, I'll remove it, good catch!

File: api/app/src/models/webhook-request.ts (hunk 4 above)
Comment (leite08) [reply api/app/src/models/webhook-request.ts:32]: removing

**Comment by Goncharo** [comment]
LGTM 🚢

**Review by Goncharo — APPROVED** [review]
🚢

File: docs/more-info/webhooks.mdx [hunk 8]
```diff
@@ -1,8 +1,157 @@
...
+## The Webhook Format
+
+The webhook format specification is as follows:
+
+1. Public endpoint accessible from the internet;
+1. Accepts a `POST` HTTP request;
+1. Verifies requests by comparing the HTTP header `x-webhook-key` with the webhook key available
+   on the [settings page](https://dash.metriport.com/settings);
+1. Accepts and responds to a [`ping` message](#the-ping-message);
+1. Accepts and processes [user data messages](#user-data-message).
```
Comment (Goncharo) [inline docs/more-info/webhooks.mdx:31]: Consistency comments:

- `1. 2. 3.`

Wording comments:

- "... on the settings page in the developer dashboard."

File: docs/more-info/webhooks.mdx [hunk 9]
```diff
@@ -1,8 +1,157 @@
...
+
+Currently, the webhook API doesn't implement retries - we're working hard to implement this soon!
```
Comment (Goncharo) [inline docs/more-info/webhooks.mdx:33]: Should we wrap this in a note/warning?

ie <Warning></Warning>

File: docs/more-info/webhooks.mdx [hunk 10]
```diff
@@ -1,8 +1,157 @@
...
+- `date-time-in-utc`: the timestamp in datetime format when this message was originally sent
+  (example: `2022-12-24T00:46:05.413Z`)
+- `users`: array of users, where each item represents all of that user's information available
+  in this request (more data might be available in subsequent requests)
+- for each user:
+  - `user-id`: the user ID as defined by your application when it called [`/create-user`](/user/create-user)
+  - `data`: array of data points for the given user; this follows the same format of the response
+    of the SDK API reference
+    - in the example above, `user-id-1` has 2 entries of [`sleep`](/health-device-api-reference/get-sleep-data)
+    data and 1 entry of [`activity`](/health-device-api-reference/get-activity-data) data
```
Comment (Goncharo) [inline docs/more-info/webhooks.mdx:154]: let's just use this to describe the response https://mintlify.com/docs/api-components/expandable , will make it cleaner

File: docs/more-info/webhooks.mdx [hunk 11]
```diff
@@ -1,8 +1,157 @@
...
+
+Upon successful receiving of these messages, it's expected that your app responds with a 200 HTTP
+status code (OK). There's no need to include anything on the response body.
```
Comment (Goncharo) [inline docs/more-info/webhooks.mdx:157]: wrap this in a note?

<Note>This adds a note in the content</Note>

**Review by Goncharo — CHANGES_REQUESTED** [review]

File: docs/more-info/webhooks.mdx (hunk 8 above)
Comment (leite08) [reply docs/more-info/webhooks.mdx:31]: > 1. 2. 3.
Markdown sets the numbering automatically, we don't need to maintain the item's numbers in order

File: docs/more-info/webhooks.mdx (hunk 11 above)
Comment (leite08) [reply docs/more-info/webhooks.mdx:157]: I can move it to a note, but its really part of the main content.

File: docs/more-info/webhooks.mdx (hunk 8 above)
Comment (Goncharo) [reply docs/more-info/webhooks.mdx]: Didn't know this!

**Review by Goncharo — APPROVED** [review]
🚢 🚢 🚢 🚢 🚢 🚢

**Pull request merged**